src/groups/mqb/mqbplug/mqbplug_authenticator.h

```diff
@@ -0,0 +1,227 @@
+#define INCLUDED_MQBPLUG_AUTHENTICATOR
+
+//@PURPOSE: Provide base classes for the 'Authenticator' plugin.
+//
+//@CLASSES:
+//  mqbplug::AuthenticationData: interface for accessing data used for
+//                               authentication.
+//  mqbplug::AuthenticationResult: the results of authentication.
+//  mqbplug::Authenticator: interface for a plugin of type 'e_AUTHENTICATOR'.
+//  mqbplug::AuthenticatorPluginFactory: base class for the 'Authenticator'
+//                                       plugin factory.
+//
+//@DESCRIPTION: This component provide definitions for classes
+// 'mqbplug::AuthenticationData', 'mqbplug::AuthenticationResult',
+// 'mqbplug::Authenticator', 'mqbplug::AuthenticatorPluginFactory' used
+// as base classes for plugins that publish statistics and for their factories.
+
+// MQB
+#include <mqbcfg_brokerconfig.h>
+#include <mqbcfg_messages.h>
+#include <mqbplug_pluginfactory.h>
+
+// BDE
+#include <bdlbb_blob.h>
+#include <bsl_iostream.h>
+#include <bsl_string.h>
+#include <bslma_managedptr.h>
+#include <bslma_usesbslmaallocator.h>
+#include <bslmf_nestedtraitdeclaration.h>
+#include <bslstl_stringref.h>
+
+namespace BloombergLP {
+namespace mqbplug {
+
+// ========================
+// class AuthenticationData
+// ========================
+
+/// Interface for accessing data used during authentication in the broker.
+class AuthenticationData {
+  public:
+    // CREATORS
+
+    /// Destroy this object.
+    virtual ~AuthenticationData();
+
+    // ACESSORS
+
+    /// Return the authentication material provided by the client.
+    virtual const bsl::vector<char>& authPayload() const = 0;
+
+    /// Return the IP Address of the client.
+    virtual bslstl::StringRef clientIpAddress() const = 0;
+};
+
+// ==========================
+// class AuthenticationResult
+// ==========================
+
+/// The results of authentication.
+class AuthenticationResult BSLS_KEYWORD_FINAL {
+  private:
+    // DATA
+
+    /// Client identity.
+    bsl::string d_principal;
+
+    /// The remaining lifetime of a client identity.
+    /// Empty value if client identity has a non-fixed lifetime.
+    bsl::optional<bsls::Types::Int64> d_lifetimeMs;
+
+  public:
+    // TRAITS
+    BSLMF_NESTED_TRAIT_DECLARATION(AuthenticationResult,
+                                   bslma::UsesBslmaAllocator)
+
+    // CREATORS
+
+    /// Construct this object using the optionally specified `allocator`.
+    explicit AuthenticationResult(bslma::Allocator* allocator = 0);
+
+    // ACCESSORS
+
+    /// Return the corresponding data field.
+    const bsl::string&                       principal() const;
+    const bsl::optional<bsls::Types::Int64>& lifetimeMs() const;
+
+    // MANIPULATORS
+    /// Reset the state of this structure.
+    void reset();
+
+    /// Set the corresponding field of this structure.
+    void setPrincipal(const bslstl::StringRef& principal);
+    void setLifetimeMs(bsls::Types::Int64 lifetimeMs);
```

Comment:
Note that with this interface we can only set empty `lifetimeMs` to a non-empty value.
We don't have a way to unset `lifetimeMs` other than calling `reset` for the whole structure.
I think we have a workflow where we need to set `lifetimeMs` and unset it right after.